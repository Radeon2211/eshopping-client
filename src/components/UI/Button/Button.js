import React from 'react';
import PropTypes from 'prop-types';
import * as SC from './Button.sc';

const Button = (props) => {
  const { clicked, children } = props;
  return (
    <SC.Button onClick={clicked} {...props}>
      {children}
    </SC.Button>
  );
};

Button.defaultProps = {
  type: 'button',
  filled: false,
  disabled: false,
  clicked: () => {},
};

Button.propTypes = {
  type: PropTypes.string,
  size: PropTypes.string.isRequired,
  filled: PropTypes.bool,
  disabled: PropTypes.bool,
  clicked: PropTypes.func,
  children: PropTypes.node.isRequired,
};

export default Button;
