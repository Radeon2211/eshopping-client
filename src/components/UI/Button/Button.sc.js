import styled from 'styled-components';

// eslint-disable-next-line import/prefer-default-export
export const Button = styled.button`
  border-radius: 1px;
  cursor: pointer;
  font-size: ${({ theme }) => theme.fontSizes.level2};
  font-weight: 700;
  letter-spacing: 2px;
  outline: none;
  padding: ${({ theme }) => theme.spacings.level1} ${({ theme }) => theme.spacings.level2};
  text-transform: uppercase;
  transition: all ${({ theme }) => theme.durations.level1}s;

  ${({ color, filled, theme }) => {
    if (color === 'blue') {
      if (filled) {
        return `
          background-color: ${theme.colors.blue};
          border: 2px solid ${theme.colors.blue};
          color: #fff;

          &:hover {
            background-color: ${theme.colors.blueLight};
            border-color: ${theme.colors.blueLight};
          }
        `;
      }
      return `
        background-color: transparent;
        border: 2px solid ${theme.colors.blue};
        color: ${theme.colors.blue};

        &:hover {
          border-color: ${theme.colors.blueLight};
          color: ${theme.colors.blueLight};
        }
      `;
    }
    if (filled) {
      return `
        background-color: ${theme.colors.red};
        border: 2px solid ${theme.colors.red};
        color: #fff;

        &:hover {
          background-color: ${theme.colors.redLight};
          border-color: ${theme.colors.redLight};
        }
      `;
    }
    return `
      background-color: transparent;
      border: 2px solid ${theme.colors.red};
      color: ${theme.colors.red};

      &:hover {
        border-color: ${theme.colors.redLight};
        color: ${theme.colors.redLight};
      }
    `;
  }}

  ${({ disabled, color, filled, theme }) => {
    if (disabled) {
      if (color === 'blue') {
        if (filled) {
          return `
            background-color: ${theme.colors.blueDark};
            border-color: ${theme.colors.blueDark};
            color: ${theme.colors.light1};
            cursor: not-allowed;

            &:hover {
              background-color: ${theme.colors.blueDark};
              border-color: ${theme.colors.blueDark};
              color: ${theme.colors.light1};
            }
          `;
        }
        return `
          border-color: ${theme.colors.blueDark};
          color: ${theme.colors.blueDark};
          cursor: not-allowed;

          &:hover {
            border-color: ${theme.colors.blueDark};
            color: ${theme.colors.blueDark};
          }
        `;
      }
      if (filled) {
        return `
          background-color: ${theme.colors.redDark};
          border-color: ${theme.colors.redDark};
          color: ${theme.colors.light1};
          cursor: not-allowed;

          &:hover {
            background-color: ${theme.colors.redDark};
            border-color: ${theme.colors.redDark};
            color: ${theme.colors.light1};
          }
        `;
      }
      return `
        border-color: ${theme.colors.redDark};
        color: ${theme.colors.redDark};
        cursor: not-allowed;

        &:hover {
          border-color: ${theme.colors.redDark};
          color: ${theme.colors.redDark};
        }
      `;
    }
    return ``;
  }}
`;
