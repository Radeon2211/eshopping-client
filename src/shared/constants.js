import * as Yup from 'yup';
import { getCountries } from 'country-fns';

export const anyStringRule = Yup.string().min(1).required();

export const userRules = {
  email: Yup.string().email().max(320).trim().required(),
  hideEmail: Yup.bool(),
  username: Yup.string().min(3).max(20).trim().required(),
  password: Yup.string().min(7).max(64).trim().required(),
  firstName: Yup.string().max(60).trim().required(),
  lastName: Yup.string().max(80).trim().required(),
  phonePrefix: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
  phoneNumber: Yup.string()
    .min(5)
    .max(15)
    .matches(/^[\d-]{5,15}$/)
    .trim()
    .required(),
  hidePhone: Yup.bool(),
  street: Yup.string().max(60).trim().required(),
  zipCode: Yup.string().max(12).trim().required(),
  city: Yup.string().max(100).trim().required(),
  country: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
};

export const productConditions = {
  NEW: 'new',
  USED: 'used',
  NOT_APPLICABLE: 'not_applicable',
};

export const productRules = {
  name: Yup.string().max(150).trim().required(),
  price: Yup.number().min(0.01).max(1000000).required(),
  quantity: Yup.number().min(1).max(100000).required(),
  condition: Yup.string()
    .oneOf([productConditions.NEW, productConditions.USED, productConditions.NOT_APPLICABLE])
    .required(),
  description: Yup.string().max(600).trim(),
};

export const listOfAreaCodes = getCountries().map(({ name, dial }) => {
  const finalValue = `+${dial} ${name.split('(')[0].trim()}`;
  return {
    value: dial,
    label: finalValue,
  };
});

export const listOfCountries = getCountries().map(({ name }) => {
  const finalValue = name.split('(')[0].trim();
  return {
    value: finalValue,
    label: finalValue,
  };
});

export const productPages = {
  ALL_PRODUCTS: 'ALL_PRODUCTS',
  MY_PRODUCTS: 'MY_PRODUCTS',
  USER_PRODUCTS: 'USER_PRODUCTS',
};

export const itemTypes = {
  CART: 'CART',
  ORDER: 'ORDER',
  TRANSACTION: 'TRANSACTION',
};

export const listItemTypes = {
  PRODUCT: 'PRODUCT',
  ORDER: 'ORDER',
};

export const orderTypes = {
  PLACED_ORDERS: 'PLACED_ORDERS',
  SELL_HISTORY: 'SELL_HISTORY',
};

export const inputKinds = {
  INPUT: 'INPUT',
  TEXTAREA: 'TEXTAREA',
  SELECT: 'SELECT',
  RADIO: 'RADIO',
};

export const defaultProductsPerPage = 10;

export const defaultOrdersPerPage = 6;

export const defaultAppPath = '/products?p=1';

export const filtersActions = {
  INIT_STATE: 'INIT_STATE',
  SET_SORT_BY: 'SET_SORT_BY',
  SET_MIN_PRICE: 'SET_MIN_PRICE',
  SET_MAX_PRICE: 'SET_MAX_PRICE',
  SET_CONDITION: 'SET_CONDITION',
};

export const sliderPositionsActions = {
  SET_BOTH: 'SET_BOTH',
  SET_LEFT: 'SET_LEFT',
  SET_RIGHT: 'SET_RIGHT',
};

export const modalTypes = {
  ABOUT_WEBSITE: 'ABOUT_WEBSITE',
  ADD_ADMIN: 'ADD_ADMIN',
  ADD_PRODUCT: 'ADD_PRODUCT',
  BUY_PRODUCTS: 'BUY_PRODUCTS',
  CART_ITEM_ADDED: 'CART_ITEM_ADDED',
  CHANGE_NAME: 'CHANGE_NAME',
  CHANGE_EMAIL: 'CHANGE_EMAIL',
  CHANGE_USERNAME: 'CHANGE_USERNAME',
  CHANGE_PHONE_NUMBER: 'CHANGE_PHONE_NUMBER',
  CHANGE_ADDRESS: 'CHANGE_ADDRESS',
  CHANGE_PASSWORD: 'CHANGE_PASSWORD',
  CHANGE_CONTACTS: 'CHANGE_CONTACTS',
  CHANGE_DELIVERY_ADDRESS: 'CHANGE_DELIVERY_ADDRESS',
  CLEAR_CART: 'CLEAR_CART',
  DELETE_ACCOUNT: 'DELETE_ACCOUNT',
  DELETE_PRODUCT: 'DELETE_PRODUCT',
  EDIT_PRODUCT: 'EDIT_PRODUCT',
  LOGIN: 'LOGIN',
  REMOVE_ADMIN: 'REMOVE_ADMIN',
  RESET_PASSWORD: 'RESET_PASSWORD',
  PENDING_USER_INFO: 'PENDING_USER_INFO',
  SEND_VERIFICATION_LINK: 'SEND_VERIFICATION_LINK',
  SIGNUP: 'SIGNUP',
};

export const sortProductsOptions = [
  { value: undefined, label: 'Default sorting' },
  { value: 'price:asc', label: `Price - ascending` },
  { value: 'price:desc', label: 'Price - descending' },
  { value: 'name:asc', label: 'Name - A to Z' },
  { value: 'name:desc', label: 'Name - Z to A' },
];

export const sortOrdersOptions = [
  { value: 'createdAt:desc', label: 'Date - from newest' },
  { value: 'createdAt:asc', label: 'Date - from oldest' },
  { value: 'overallPrice:asc', label: `Price - ascending` },
  { value: 'overallPrice:desc', label: 'Price - descending' },
];

export const singleInfoNames = {
  USERNAME: 'Username',
  NAME: 'Name',
  EMAIL: 'Email',
  PHONE_NUMBER: 'Phone number',
  ADDRESS: 'Address',
  CONTACTS: 'Contacts',
};

export const updateCartActions = {
  INCREMENT: 'INCREMENT',
  DECREMENT: 'DECREMENT',
  NUMBER: 'NUMBER',
};

export const defaultErrorMessage = 'Something went wrong';

export const productsPerPageControllerOptions = [
  { value: 1, label: '1' },
  { value: 2, label: '2' },
  { value: 3, label: '3' },
  { value: 4, label: '4' },
  { value: 5, label: '5' },
  { value: 10, label: '10' },
  { value: 15, label: '15' },
  { value: 20, label: '20' },
  { value: 25, label: '25' },
];

export const defaultScrollToConfig = {
  top: 0,
  left: 0,
  behavior: 'smooth',
};

export const userStatuses = {
  ACTIVE: 'active',
  PENDING: 'pending',
};
