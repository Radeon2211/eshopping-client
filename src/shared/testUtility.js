// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const historyPageNum = (pageNumber) => ({
  listen: jest.fn(),
  location: { search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push: jest.fn(),
});

export const propsPagination = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  maxQuantityPerPage: 2,
});
