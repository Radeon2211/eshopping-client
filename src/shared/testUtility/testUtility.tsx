/* eslint-disable import/no-extraneous-dependencies */
import { ReactElement } from 'react';
import { Router } from 'react-router-dom';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import { MockStoreEnhanced } from 'redux-mock-store';
import { fireEvent, waitFor, render } from '@testing-library/react';
import { v4 as uuidv4 } from 'uuid';
import moxios from 'moxios';
import rootReducer from '../../store/reducers/rootReducer';
import theme from '../../styled/theme';
import {
  CartItem,
  DeliveryAddress,
  Order,
  OrderProductItem,
  Product,
  Profile,
  ProfileStatus,
} from '../types/types';
import { ProductCondition } from '../types/enums';

export const defaultDeliveryAddress: DeliveryAddress = {
  firstName: 'firstName',
  lastName: 'lastName',
  street: 'street',
  zipCode: 'zipCode',
  city: 'city',
  country: 'country',
  phone: 'phone',
};

export const defaultUserProfile: Profile = {
  ...defaultDeliveryAddress,
  _id: 'userId',
  username: 'username',
  email: 'email@domain.com',
  status: ProfileStatus.ACTIVE,
  contacts: {
    email: true,
    phone: true,
  },
  createdAt: '2021-02-10T19:10:38.872Z',
};

export const createPaginationProps = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  quantityPerPage: 2,
});

export const createProductItem = (props: Partial<Product> = {}): Product => {
  return {
    _id: uuidv4(),
    quantity: 1,
    price: 2,
    name: 'product name',
    photo: false,
    quantitySold: 0,
    buyerQuantity: 0,
    description: '',
    condition: ProductCondition.NEW,
    seller: {
      username: 'username',
    },
    createdAt: '2021-02-10T19:10:38.872Z',
    updatedAt: '2021-02-10T19:10:38.872Z',
    __v: 0,
    ...props,
  };
};

export const createCartItem = (props = {}): CartItem => {
  const finalProps = {
    id: uuidv4(),
    sellerUsername: 'username',
    quantity: 1,
    productId: uuidv4(),
    price: 2,
    name: 'product name',
    productQuantity: 2,
    photo: false,
    ...props,
  };
  const product = createProductItem({
    _id: finalProps.productId,
    quantity: finalProps.productQuantity,
    price: finalProps.price,
    name: finalProps.name,
    photo: finalProps.photo,
    seller: {
      username: finalProps.sellerUsername,
    },
  });

  return {
    _id: finalProps.id,
    quantity: finalProps.quantity,
    product,
  };
};

export const createTransactionAndOrderProdItem = (
  props: Partial<OrderProductItem> = {},
): OrderProductItem => {
  return {
    _id: uuidv4(),
    quantity: 1,
    price: 2,
    name: 'product name',
    photo: false,
    seller: {
      username: 'username',
    },
    ...props,
  };
};

export const createOrder = (props: Partial<Order> = {}): Order => {
  return {
    _id: uuidv4(),
    overallPrice: 100,
    seller: null,
    buyer: null,
    products: [createTransactionAndOrderProdItem()],
    deliveryAddress: defaultDeliveryAddress,
    createdAt: '2021-02-10T19:10:38.872Z',
    updatedAt: '2021-02-10T19:10:38.872Z',
    __v: 0,
    ...props,
  };
};

export const testStore = (initialAuth = {}, initialProduct = {}, initialUI = {}) => {
  const createStoreWithMiddleware = applyMiddleware(thunk)(createStore);
  const defaultState = createStoreWithMiddleware(rootReducer).getState();

  const finalState = {
    auth: {
      ...defaultState.auth,
      ...initialAuth,
    },
    product: {
      ...(defaultState.product as object),
      ...initialProduct,
    },
    ui: {
      ...(defaultState.ui as object),
      ...initialUI,
    },
  };
  // @ts-expect-error TBF after redux upgrade
  const finalStore = createStoreWithMiddleware(rootReducer, finalState);

  return {
    store: finalStore,
    initialState: finalStore.getState(),
  };
};

type MoxiosRequestsItems = { config: { method: string; url: string } }[];
type MockedMoxiosInstance = {
  requests: {
    __items?: MoxiosRequestsItems;
    mostRecent?: () => { config: { method: string; url: string } };
  };
};
type DefaultMoxiosInstance = Omit<typeof moxios, 'requests'> & {
  requests: typeof moxios.requests & {
    __items?: MoxiosRequestsItems;
  };
};
export const checkReqMethodAndURL = (
  moxiosInstance: MockedMoxiosInstance | DefaultMoxiosInstance,
  method: string,
  url: string,
  reqNumber = 0,
) => {
  if (reqNumber) {
    const requests = moxiosInstance.requests.__items;
    const prevRequest = requests?.[reqNumber - 1];
    return prevRequest && method === prevRequest.config.method && url === prevRequest.config.url;
  }
  if (!moxiosInstance.requests.mostRecent) return false;
  const lastRequest = moxiosInstance.requests.mostRecent();
  return method === lastRequest.config.method && url === lastRequest.config.url;
};

export const createExpectedState = (
  initialState: {
    auth?: object;
    product?: object;
    ui?: object;
  },
  expectedAuth = {},
  expectedProduct = {},
  expectedUI = {},
) => {
  const newState = {
    auth: {
      ...initialState.auth,
      ...expectedAuth,
    },
    product: {
      ...initialState.product,
      ...expectedProduct,
    },
    ui: {
      ...initialState.ui,
      ...expectedUI,
    },
  };

  return newState;
};

export const setUpStoreWithDefaultProfile = (auth = {}, product = {}, ui = {}) => {
  return testStore(
    {
      profile: defaultUserProfile,
      deliveryAddress: defaultDeliveryAddress,
      cart: [],
      ...auth,
    },
    product,
    ui,
  );
};

export const clickAtSubmitButton = async (container: Element) => {
  await waitFor(() => {
    const submitBtn = container.querySelector('button[type="submit"]');
    if (submitBtn) {
      fireEvent.click(submitBtn);
    }
  });
};

export const testRouterPushCall = (
  pushFn: ReturnType<typeof jest.fn>,
  callIdx: number,
  expectedPathname: string,
  expectedSearch = '',
) => {
  expect(pushFn.mock.calls[callIdx][0].pathname).toEqual(expectedPathname);
  expect(pushFn.mock.calls[callIdx][0].search).toEqual(expectedSearch);
};

type RenderAppPartOptions = {
  push?: ReturnType<typeof jest.fn>;
  pathname?: string;
  search?: string;
  withoutThemeProvider?: boolean;
  withoutRouter?: boolean;
  store?: MockStoreEnhanced<object, object>;
  onlyReturnWrappedElement?: boolean;
};

export const renderAppPart = (element: ReactElement, options: RenderAppPartOptions = {}) => {
  const navigator = {
    createHref: jest.fn(),
    push: options.push || jest.fn(),
    go: jest.fn(),
    replace: jest.fn(),
  };

  const location = {
    pathname: options.pathname || '/',
    search: options.search || '',
  };

  let content = element;
  if (!options.withoutThemeProvider) {
    content = <ThemeProvider theme={theme}>{element}</ThemeProvider>;
  }
  if (options.store) {
    content = <Provider store={options.store}>{content}</Provider>;
  }
  if (!options.withoutRouter) {
    content = (
      <Router location={location} navigator={navigator}>
        {content}
      </Router>
    );
  }

  return options.onlyReturnWrappedElement ? content : render(content);
};
