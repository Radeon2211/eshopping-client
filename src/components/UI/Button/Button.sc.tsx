import styled from 'styled-components';
import { ButtonProps } from './Button';

// eslint-disable-next-line import/prefer-default-export
export const Button = styled.button<{
  $size: ButtonProps['$size'];
  $stretch: ButtonProps['$stretch'];
  $filled: ButtonProps['$filled'];
  $color: ButtonProps['$color'];
}>`
  box-shadow: ${({ theme }) => theme.shadows.level1};
  border-radius: 1px;
  cursor: pointer;
  font-weight: 700;
  line-height: ${({ theme }) => theme.lineHeights.level3};
  letter-spacing: 2px;
  outline: none;
  position: relative;
  text-align: center;
  text-transform: uppercase;
  transition: all ${({ theme }) => theme.durations.level1}s;

  ${({ $size, theme }) => {
    if ($size === 'big') {
      return `
        font-size: ${theme.fontSizes.level4};
        padding: ${theme.spacings.level2} ${theme.spacings.level3};
      `;
    }
    return `
    font-size: ${theme.fontSizes.level2};
        padding: ${theme.spacings.level1} ${theme.spacings.level2};
    `;
  }}

  ${({ $stretch }) => {
    return $stretch ? `width: 100%;` : ``;
  }}

  ${({ $color, $filled, theme }) => {
    if ($color === 'blue') {
      if ($filled) {
        return `
          background-color: ${theme.colors.blue};
          border: 2px solid ${theme.colors.blue};
          color: #fff;

          &:hover {
            background-color: ${theme.colors.blueLight};
            border-color: ${theme.colors.blueLight};
          }
        `;
      }
      return `
        background-color: transparent;
        border: 2px solid ${theme.colors.blue};
        color: ${theme.colors.blue};

        &:hover {
          border-color: ${theme.colors.blueLight};
          color: ${theme.colors.blueLight};
        }
      `;
    }
    if ($filled) {
      return `
        background-color: ${theme.colors.red};
        border: 2px solid ${theme.colors.red};
        color: #fff;

        &:hover {
          background-color: ${theme.colors.redLight};
          border-color: ${theme.colors.redLight};
        }
      `;
    }
    return `
      background-color: transparent;
      border: 2px solid ${theme.colors.red};
      color: ${theme.colors.red};

      &:hover {
        border-color: ${theme.colors.redLight};
        color: ${theme.colors.redLight};
      }
    `;
  }}

  ${({ disabled, $color, $filled, theme }) => {
    if (disabled) {
      if ($color === 'blue') {
        if ($filled) {
          return `
            background-color: ${theme.colors.blueDark};
            border-color: ${theme.colors.blueDark};
            color: ${theme.colors.light1};
            cursor: not-allowed;

            &:hover {
              background-color: ${theme.colors.blueDark};
              border-color: ${theme.colors.blueDark};
              color: ${theme.colors.light1};
            }
          `;
        }
        return `
          border-color: ${theme.colors.blueDark};
          color: ${theme.colors.blueDark};
          cursor: not-allowed;

          &:hover {
            border-color: ${theme.colors.blueDark};
            color: ${theme.colors.blueDark};
          }
        `;
      }
      if ($filled) {
        return `
          background-color: ${theme.colors.redDark};
          border-color: ${theme.colors.redDark};
          color: ${theme.colors.light1};
          cursor: not-allowed;

          &:hover {
            background-color: ${theme.colors.redDark};
            border-color: ${theme.colors.redDark};
            color: ${theme.colors.light1};
          }
        `;
      }
      return `
        border-color: ${theme.colors.redDark};
        color: ${theme.colors.redDark};
        cursor: not-allowed;

        &:hover {
          border-color: ${theme.colors.redDark};
          color: ${theme.colors.redDark};
        }
      `;
    }
    return ``;
  }}
`;
