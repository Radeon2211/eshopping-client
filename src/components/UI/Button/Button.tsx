import * as SC from './Button.sc';
import LoadingOverlay from '../LoadingOverlay/LoadingOverlay';
import { PropsWithChildren } from '../../../shared/types/types';

export interface ButtonProps {
  $size?: 'small' | 'big';
  type?: 'submit' | 'reset' | 'button';
  $color?: 'blue' | 'red';
  $filled?: boolean;
  $stretch?: boolean;
  isLoading?: boolean;
  disabled?: boolean;
  clicked?: () => void;
}

export default function Button(props: PropsWithChildren<ButtonProps>) {
  const {
    clicked = () => {},
    isLoading = false,
    disabled = false,
    type = 'button',
    $size = 'small',
    $color = 'blue',
    $stretch = false,
    $filled = false,
    children,
    ...rest
  } = props;

  const loadingOverlay = isLoading ? <LoadingOverlay loaderSize="small" disableText /> : null;

  return (
    <SC.Button
      onClick={clicked}
      disabled={disabled || isLoading}
      type={type}
      $size={$size}
      $color={$color}
      $stretch={$stretch}
      $filled={$filled}
      {...rest}
    >
      {children}
      {loadingOverlay}
    </SC.Button>
  );
}
