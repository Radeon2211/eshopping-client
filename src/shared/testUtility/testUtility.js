// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';
import { v4 as uuidv4 } from 'uuid';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from '../../store/reducers/rootReducer';
import { productConditions } from '../constants';

export const defaultDeliveryAddress = {
  firstName: 'firstName',
  lastName: 'lastName',
  street: 'street',
  zipCode: 'zipCode',
  city: 'city',
  country: 'country',
  phone: 'phone',
};

export const defaultUserProfile = {
  ...defaultDeliveryAddress,
  _id: 'userId',
  username: 'username',
  email: 'email@domain.com',
  status: 'active',
  contacts: {
    email: true,
    phone: true,
  },
  createdAt: '2021-02-10T19:10:38.872Z',
};

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const createHistoryPageNumber = (pageNumber = 1, push = jest.fn()) => ({
  listen: jest.fn(),
  location: { pathname: '/products', search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push,
});

export const createPaginationProps = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  quantityPerPage: 2,
});

export const createProductItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
  quantitySold = 0,
  buyerQuantity = 0,
  description = '',
  condition = productConditions.NEW,
) => ({
  _id: productId || uuidv4(),
  name,
  price,
  quantity,
  quantitySold,
  buyerQuantity,
  description,
  condition,
  photo,
  seller: {
    username: sellerUsername,
  },
  __v: 0,
  createdAt: '2021-02-10T19:10:38.872Z',
  updatedAt: '2021-02-10T19:10:38.872Z',
});

export const createCartItem = (
  sellerUsername = 'username',
  quantity = 1,
  productId,
  price = 2,
  name = 'product name',
  productQuantity = 2,
  photo = false,
) => {
  const product = createProductItem(
    productId || uuidv4(),
    sellerUsername,
    productQuantity,
    price,
    name,
    photo,
  );

  return {
    _id: uuidv4(),
    quantity,
    product,
  };
};

export const createTransactionAndOrderProdItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
) => ({
  _id: productId || uuidv4(),
  name,
  photo,
  price,
  quantity,
  seller: {
    username: sellerUsername,
  },
});

export const createOrder = (
  products = [createTransactionAndOrderProdItem()],
  orderId = uuidv4(),
  sellerUsername,
  buyerUsername,
  overallPrice,
  createdAt,
  sellerEmail,
  sellerPhone,
) => {
  const seller = sellerUsername
    ? {
        username: sellerUsername,
        email: sellerEmail,
        phone: sellerPhone,
      }
    : null;

  const buyer = buyerUsername
    ? {
        username: buyerUsername,
      }
    : null;

  return {
    _id: orderId,
    seller,
    buyer,
    products,
    overallPrice,
    deliveryAddress: defaultDeliveryAddress,
    createdAt,
    updatedAt: createdAt,
    __v: 0,
  };
};

export const testStore = (initialAuth = {}, initialProduct = {}, initialUI = {}) => {
  const createStoreWithMiddleware = applyMiddleware(thunk)(createStore);
  const defaultState = createStoreWithMiddleware(rootReducer).getState();

  const finalState = {
    auth: {
      ...defaultState.auth,
      ...initialAuth,
    },
    product: {
      ...defaultState.product,
      ...initialProduct,
    },
    ui: {
      ...defaultState.ui,
      ...initialUI,
    },
  };
  const finalStore = createStoreWithMiddleware(rootReducer, finalState);

  return {
    store: finalStore,
    initialState: finalStore.getState(),
  };
};

export const checkReqMethodAndURL = (moxiosInstance, method, url, reqNumber = 0) => {
  if (reqNumber) {
    const requests = moxiosInstance.requests.__items;
    const prevRequest = requests[reqNumber - 1];
    return method === prevRequest.config.method && url === prevRequest.config.url;
  }
  const lastRequest = moxiosInstance.requests.mostRecent();
  return method === lastRequest.config.method && url === lastRequest.config.url;
};

export const createExpectedState = (
  initialState,
  expectedAuth = {},
  expectedProduct = {},
  expectedUI = {},
) => {
  const newState = {
    auth: {
      ...initialState.auth,
      ...expectedAuth,
    },
    product: {
      ...initialState.product,
      ...expectedProduct,
    },
    ui: {
      ...initialState.ui,
      ...expectedUI,
    },
  };

  return newState;
};

export const setUpStoreWithDefaultProfile = (auth = {}, product = {}, ui = {}) => {
  return testStore(
    {
      profile: defaultUserProfile,
      deliveryAddress: defaultDeliveryAddress,
      cart: [],
      ...auth,
    },
    product,
    ui,
  );
};
