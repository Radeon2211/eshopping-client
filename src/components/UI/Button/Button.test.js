import React from 'react';
import { shallow } from 'enzyme';
import Button from './Button';
import * as SC from './Button.sc';
import { checkProps } from '../../../shared/testUtility';

const setUp = (props = {}) => {
  return shallow(<Button {...props} />);
};

describe('<Button />', () => {
  describe('Check prop types', () => {
    const expectedProps = {
      size: 'small',
      children: 'test button',
    };

    it('Should NOT throw a warning', () => {
      expect(checkProps(Button, expectedProps)).toBeUndefined();
    });

    it('Should throw a warning', () => {
      expect(checkProps(Button, {})).not.toBe(null);
    });
  });

  describe('Check if button renders', () => {
    let wrapper;
    let clicked;
    beforeEach(() => {
      clicked = jest.fn();
      const props = {
        size: 'small',
        children: 'test button',
        clicked,
      };
      wrapper = setUp(props);
    });

    it('Should render a button', () => {
      expect(wrapper.find(SC.Button)).toHaveLength(1);
    });

    it('Should emit callback on click event', () => {
      const button = wrapper.find(SC.Button);
      button.simulate('click');
      expect(clicked).toHaveBeenCalled();
    });
  });
});
