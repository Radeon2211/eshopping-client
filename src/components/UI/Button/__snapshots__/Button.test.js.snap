// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Check how renders Should render with <LoadingOverlay /> 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ clkCDw"
    color="blue"
    disabled=""
    type="button"
  >
    Test button
    <div
      class="sc-eCssSg ida-dSY"
      data-testid="LoadingOverlay"
    >
      <div
        class="sc-dlfnbm fSwSVd"
        data-testid="Loader"
      >
        <div />
        <div />
        <div />
        <div />
      </div>
    </div>
  </button>
</DocumentFragment>
`;

exports[`<Button /> Check how renders Should render without <LoadingOverlay /> 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ eqGxsp"
    color="blue"
    type="button"
  >
    Test button
  </button>
</DocumentFragment>
`;
